tensorrt_llm/_torch/models/modeling_llava_next.py

```diff
@@ -51,6 +51,195 @@ def __init__(self,
 
         self.image_token_index = model_config.image_token_index
         self.vocab_size = model_config.vocab_size
+        self.vision_feature_select_strategy = getattr(
+            model_config, "vision_feature_select_strategy",
+            "default")
+        self.config = model_config.vision_config
+
+    def _get_num_unpadded_features(
+        self,
+        *,
+        original_height: int,
+        original_width: int,
+        npatches: int,
+        num_patch_height: int,
+        num_patch_width: int,
+    ) -> tuple[int, int]:
+        current_height = npatches * num_patch_height
+        current_width = npatches * num_patch_width
+
+        aspect_ratio = original_width / original_height
+        current_aspect_ratio = current_width / current_height
+
+        if aspect_ratio > current_aspect_ratio:
+            new_height = int(
+                round(original_height * (current_width / original_width), 7))
+            padding = (current_height - new_height) // 2
+            current_height = current_height - (2 * padding)
+        else:
+            new_width = int(
+                round(original_width * (current_height / original_height), 7))
+            padding = (current_width - new_width) // 2
+            current_width = current_width - (2 * padding)
+
+        unpadded_features = current_height * current_width
+        newline_features = current_height
+
+        return (unpadded_features, newline_features)
+
+    # Based on: https://github.com/huggingface/text-generation-inference/blob/v3.0.1/server/text_generation_server/models/vlm_causal_lm.py#L113
+    def get_num_tokens_per_image(
+        self,
+        *,
+        image_width: int,
+        image_height: int,
+    ) -> int:
+        patch_grid_length = self.config.image_size // self.config.patch_size
+        base_feature_size = patch_grid_length**2 + 1
+
+        if self.vision_feature_select_strategy == "default":
+            base_feature_size = base_feature_size - 1
+
+        num_patch_height, num_patch_width = get_anyres_image_grid_shape(
+            image_size=(image_width, image_height),
+            grid_pinpoints=self.model_config.image_grid_pinpoints,
+            patch_size=self.config.image_size,
+        )
+
+        (
+            unpadded_feature_size,
+            newline_feature_size,
+        ) = self._get_num_unpadded_features(
+            original_height=image_height,
+            original_width=image_width,
+            npatches=patch_grid_length,
+            num_patch_height=num_patch_height,
+            num_patch_width=num_patch_width,
+        )
+        return unpadded_feature_size + newline_feature_size + base_feature_size
+
+    def _postprocess(self, input_ids, mm_features):
```

Comment:
Can we use type hints here too?

Reply:
Note: original functions are removed for simplicity